src/main/scala/zio/prelude/Idempotent.scala

```diff
@@ -0,0 +1,498 @@
+package zio.prelude
+
+import zio.prelude.coherent.EqualIdempotent
+import zio.test.TestResult
+import zio.test.laws.{ Lawful, Laws }
+
+/**
+ * The `Idempotent` type class describes a binary operator for a type `A` that
+ * is both associative and produces the same value when combining two identical values.
+ * This means that `a <> a` is equal to `a` for all values `a`.
+ * Example of idempotent operations is union of sets, but not addition of integers.
+ *
+ * Idempotent operators are useful because combining  the values with an idempotent
+ * operation results in the same value regardless of the number of values
+ * are combined, allowing us to optimize out unnecessary combinations of the same values.
+ */
+trait Idempotent[A] extends Associative[A] { self =>
+
+  protected def combineNormal(l: => A, r: => A): A = combine(l, r)
+
+  final def combineIdempotent(l: => A, r: => A)(implicit A: Equal[A]): A =
+    if (l === r) l else combineNormal(l, r)
+
+  def idempotent(implicit A: Equal[A]): Idempotent[A] = new Idempotent[A] {
```

Comment:
Isn't this more a constructor that should be in the companion object rather than a method on the `Idempotent` trait. Like `fromEqual`?

Reply:
I took an inspiration from `Commutative`, which has `commute`.